File: Coading Tips And Feature/Learning a Typescript/Data_Type.ts
// https://chatgpt.com/c/684cfdc6-75a0-8011-9a0e-301de4f7006a

console.log("This is a Data Type file");

// Data types in TypeScript

// 1) Primitive Data Types 

// 1. Number
let num: number = 42;
// console.log("Number:", num);


// 2. String
let str: string = "Hello, TypeScript!";
// console.log("String:", str);

// 3. Boolean
let Bol: boolean = true;
// console.log("Boolean:", Bol);


// 4. Bigint
let bignt: bigint = 90071992549999999999n; // Note the 'n' at the end to denote BigInt
// console.log("BigInt:", bignt);

// 5. nuull
let nul: null = null;
// console.log("Null:", nul);

// 6. undefined
let undef: undefined = undefined;
// console.log("Undefined:", undef);


// Non-Primitive Data Types

// 1. Object
let person: { name: string, age: number, isStudent: boolean, phone: number } = {
    name: "John Doe",
    age: 30,
    isStudent: false,
    phone: 1234567890
}
// console.log("Object:", person);

// 2. Array
let arr: number[] = [1, 2, 3, 4, 5];
let arr_1: string[] = ["apple", "banana", "cherry"];
// console.table(`"Array of Strings:" ${arr_1} "Array of Number:" ${arr} `);

// 3. Function
function john(a: number, b: number): number {
    return a + b;
}
// console.log("Function Result:", john(5, 10));

// 4. Tuple
let tuple: [string, number, boolean] = ["Alice", 25, true];
// console.log("Tuple:", tuple);

// 5. Class

class Person {
    name: string;
    age: number;
    isStudent: boolean;
    phone: number;
    constructor(name: string, age: number, isStudent: boolean, phone: number) {
        this.name = name;
        this.age = age;
        this.isStudent = isStudent;
        this.phone = phone;
    }
}

const check_person = new Person("Tanver", 25, true, 1234567890);
// console.log("Class:", check_person);

// 7. Interface

interface first_person {
    name: string;
    age: number;
    isstudent: boolean;
    phone: number;
    attendance: boolean;
    address: string;
    weight: number;
}

const data_person: first_person = {
    name: "Mukesh",
    age: 19,
    phone: 1234567890,
    isstudent: true,
    address: "123 Main St",
    attendance: true,
    weight: 70
}
// console.log(data_person);


//  3. Special & Advanced Types

// 1. enum (enumeration)

enum Color {
    Red = "RED",
    Green = "GREEN",
    Blue = "BLUE"
}
// console.log("Enum Color:", Color.Red);

// 2. any

// any ایک ایسا type ہے جو کسی بھی طرح کی value کو accept کرتا ہے، یعنی:

// string, number, boolean, array, object، کچھ بھی assign کر سکتے ہیں۔

// TypeScript کی type-checking temporarily disable ہو جاتی ہے۔

let anything: any = "This can be anything"; // Now it's a string
anything = 42; // Now it's a number
// console.log("Any Type:", anything);

// 3. unknown
let value: unknown = "This can be anything too";

if (typeof value === "string") {
    // console.log("Value is a string:", value.toUpperCase());
    // console.log(typeof value);
}

// 4. never
// agr koi bhi asa function hai jo kabhi bhi return nahi hoti, ya kabhi bhi terminate nahi hoti, to usay never type kehte hain. or agr infinte ho to bhi is ko never khta hai

function infi(): never {
   while (true) {
    console.log("This will run forever");
    
   }
}
// console.log(infi());


// 5. void
// hum agr koi bhi value return nahi karte, ya function ka koi return type nahi hota, to hum void type use karte hain

function viod(name:string): void{
    console.log("This function does not return anything:", name);   
}
// viod("Bilal");

// 6. function types

let check_fun:(a:number, b:number) => number;
check_fun = function(a,b) {
    return a + b;
}
// console.log(check_fun(10 , 32));

// 7. unoin type

// ya hma sara dta tpe acces krna ka dta hai ak line ma hum is ma likh da ka string nubmer boolean to wo in sab ko acces kr lata hai
let All_access : string | number | boolean | number[] | string[];

All_access = `This can be a string, number, boolean, or array.`;
// console.log("Union Type In String:", All_access);
All_access = 42;
// console.log("Union Type In Number:", All_access);
All_access = true;
// console.log("Union Type In Boolean:", All_access);
All_access = [1, 2, 3, 4, 5];
// console.log("Union Type In Array:", All_access);
All_access = ["apple", "banana", "cherry"];
// console.log("Union Type In Array:", All_access);

// 8. intersection type
// Intersection type ka matlab hai ke hum do ya zyada types ko combine karte hain, yani ke dono types ki properties ko ek sath use karte hain

type A = {name : string}
type B = {age : number}

type C = A & B; // Intersection type

const persondet:C ={
    name: "John",
    age: 30
}
// console.log("Intersection :", persondet);

// 9. literal type
// Literal type ka matlab hai ke hum kisi variable ko sirf ek specific value assign karte hain, yani ke wo sirf usi value ko hold kar sakta hai

let direction: "left" | "right" | "up" | "down" | "backward";

direction = "left"; // Valid
direction = "right"; // Valid
direction = "up"; // Valid
direction = "down"; // Valid
// direction = "forward"; // Invalid, will cause a TypeScript error

// console.log("Direction is Literal:", direction);

// 10. type aliases
// Type aliases ka matlab hai ke hum kisi type ko ek naya naam de dete hain, yani ke hum us type ko as a new type use kar sakte hain

type channel = {
    name: string;
    subscribers: number;
    isLive: boolean;
}
const youtubeChannel: channel = {
    name: "Tech Channel",
    subscribers: 1000,
    isLive: true
};
// console.log(youtubeChannel);


// 11. Generic types
// Generic types ka matlab hai ke hum kisi type ko generic bana dete hain, yani ke wo kisi bhi type ko accept kar sakta hai
function identitity<T>(arg: T):T{
    return arg;
}
console.log("Generic Type:", identitity<string>("Hello, Generic!"));

// ***************** Complete a Data Type *****************

File: Coading Tips And Feature/Learning a Typescript/Data_Type.py
from fpdf import FPDF

# PDF Setup
pdf = FPDF()
pdf.add_page()
pdf.set_auto_page_break(auto=True, margin=15)
pdf.set_font("Arial", size=12)

# Title
pdf.set_font("Arial", 'B', 16)
pdf.cell(0, 10, "TypeScript Data Types (With Examples)", ln=True, align='C')
pdf.ln(10)
pdf.set_font("Arial", size=12)

# Content Sections
content = [
    ("1. Primitive Data Types", """
Primitive types are built-in and immutable:

- number: 10, 3.14, -5
- string: "Hello", 'TS'
- boolean: true, false
- null: null
- undefined: undefined
- bigint: 123456789n
- symbol: Symbol("id")
    
Examples:
let age: number = 25;
let username: string = "Tanveer";
let isLoggedIn: boolean = true;
let salary: bigint = 10000000000n;
let id: symbol = Symbol("userId");
let nothing: null = null;
let notAssigned: undefined = undefined;
"""),

    ("2. Non-Primitive (Reference) Data Types", """
Non-primitive types are mutable and stored by reference:

- object: { name: "Ali", age: 30 }
- array: [10, 20, 30]
- tuple: ["Tanveer", 21]
- function: function greet() {}
- class: class Student {}
- enum: enum Direction { Up, Down }

Examples:
let person: { name: string; age: number } = { name: "Ali", age: 30 };
let scores: number[] = [10, 20, 30];
let user: [string, number] = ["Tanveer", 21];
function greet(): void { console.log("Hello, TypeScript"); }
class Student { name: string; constructor(name: string) { this.name = name; } }
enum Direction { Up, Down, Left, Right }
"""),

    ("3. Special Data Types", """
Used in special scenarios:

- any: Accepts any value (no type checking)
- unknown: Like any, but safer
- never: Function that never returns
- void: Function with no return value

Examples:
let random: any = "hello"; random = 10;
let input: unknown = "TS";
if (typeof input === "string") { console.log(input.toUpperCase()); }
function throwError(): never { throw new Error("Error occurred"); }
function logMessage(): void { console.log("This function returns nothing"); }
"""),

    ("4. Basic TypeScript Code Summary", """
// Primitive Types
let age: number = 20;
let name: string = "Ali";

// Non-Primitive
let user: { id: number, name: string } = { id: 1, name: "Tanveer" };
let arr: string[] = ["a", "b", "c"];
let tup: [number, boolean] = [1, true];

// Special
let anything: any = 10;
let unkn: unknown = "text";
function test(): void {}
function crash(): never { throw new Error("fail"); }
"""),

    ("Conclusion", """
Category        | Included Types
----------------|----------------------------------------------------------
Primitive       | number, string, boolean, null, undefined, symbol, bigint
Non-Primitive   | object, array, tuple, function, class, enum
Special         | any, unknown, never, void
""")
]

# Add content to PDF
for heading, text in content:
    pdf.set_font("Arial", 'B', 14)
    pdf.cell(0, 10, heading, ln=True)
    pdf.set_font("Arial", size=12)
    for line in text.strip().split('\n'):
        pdf.multi_cell(0, 8, line)
    pdf.ln(3)

# Save PDF
pdf_path = "/mnt/data/TypeScript_Data_Types_Complete_Guide.pdf"
pdf.output(pdf_path)
